src/jit/morph.cpp

```diff
@@ -17162,6 +17162,12 @@ void Compiler::fgMorphStructField(GenTree* tree, GenTree* parent)
                     varDsc->incLvRefCnt(1, RCS_EARLY);
                 }
 
+                if (genTypeSize(tree->gtType) != genTypeSize(lvaTable[fieldLclIndex].TypeGet()))
+                {
```

Comment:
We can allow smaller sized access here.
It is only when we are reading a larger size that we have the problem:

Reply:
Yes, I think it is safe, but I think that others part of Jit do not expect such conversions.
We will have smth like:
```
[000601] ------------              *  STMT      void  (IL   ???...  ???)
[000854] ------------              |  /--*  LCL_VAR   long   (AX) V46 tmp45
[000600] -AC---------              \--*  ASG       int  
[000599] D------N----                 \--*  LCL_VAR   int V30 tmp29 
```
that doesn't look right, does it?

Reply:
I think that allowing a smaller sized access would require a cast. One would have to deduce whether we're reading or writing, and then cast accordingly. I'm not sure it's worth the extra complexity, but a comment indicating that it would be safe but not worth doing would be good.
That said, I believe that, at this point, the decision to promote has already been made. If we decide not to make the transformation, we need to set the variable as `lvDoNotEnregister` (presumably `DNER_LocalField`).

Reply:
I haven't been able to reproduce the issue so perhaps I'm missing something. But it seems to me that we should check the type (the "actual" type really) and not their sizes. Otherwise the same issue will likely show up for float/int and double/long "reinterpretation".

That probably can be made to work by inserting a bitcast node but this is probably not the right time to attempt such optimizations. And anyway this code should be rewritten and moved to LocalAddressVisitor. As is now it kind of duplicates logic done by that.

Reply:
To repro the issue cherry-pick this change: https://github.com/dotnet/coreclr/commit/d485659348af3528b688ac5b7e4c58dacba6de58.

Reply:
I tried to make R2R corelib using the ReadyToRun crossgen parameter, perhaps it doesn't have the same effect. Anyway, too much trouble to reproduce using the existing test. Speaking of floating/integer mismatches here's a test that produces an assert:
```C#


```

Reply:
Good ol' x64 non-tiered produces
```
Assert failure(PID 11844 [0x00002e44], Thread: 13556 [0x34f4]): Assertion failed '!varTypeIsFloating(targetType) || (targetType == data->TypeGet())' in 'Program2:Test(struct)' (IL size 54)

    File: d:\projects\coreclr\src\jit\codegenxarch.cpp Line: 4675
    Image: D:\Projects\coreclr\bin\Product\Windows_NT.x64.Checked\CoreRun.exe
```